File: 0019_Remove_Nth_Node_From_End_of_List/solution.py
'''
19. Remove Nth Node From End of List

Given a linked list, remove the nth node from the end of list and return its head.

For example,

   Given linked list: 1->2->3->4->5, and n = 2.

   After removing the second node from the end, the linked list becomes 1->2->3->5.

'''
from data_structure.list_node import ListNode


class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        feak_head = ListNode(0)
        feak_head.next = head
        slow = feak_head
        fast = head
        for i in range(0, n):
            fast = fast.next

        while fast != None:
            fast = fast.next
            slow = slow.next
        slow.next = slow.next.next
        return feak_head.next

File: 0019_Remove_Nth_Node_From_End_of_List/solution.cpp
/*
19. Remove Nth Node From End of List

Level: Medium

https://leetcode.com/problems/remove-nth-node-from-end-of-list
*/
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>

#include <algorithm>

#include "common/ListNode.h"
#include "common/TreeNode.h"

using namespace std;

/*
  Solution: 
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
  ListNode *removeNthFromEnd(ListNode *head, int n)
  {
    ListNode *dummy = new ListNode(0);
    dummy->next = head;
    ListNode *start = dummy;
    ListNode *end = dummy;

    // Note "start" is node before Nth node from end 
    while (n >= 0)
    {
      end = end->next;
      n--;
    }

    while (end)
    {
      end = end->next;
      start = start->next;
    }
    start->next = start->next->next;
    return dummy->next;
  }
};